cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaDistributionTarget.java

```diff
@@ -108,30 +108,27 @@ boolean moveReplicasInSourceBrokerToEligibleBrokers(ClusterModel clusterModel,
   }
 
   /**
-   * Move given self healing eligible replica residing in the given cluster in a dead or healthy broker to an eligible
+   * Move given self healing eligible replica residing in the given cluster in a dead or alive broker to an eligible
    * broker. Replica movements are guaranteed not to violate the requirements of optimized goals.
    *
    * @param clusterModel     The state of the cluster.
    * @param replicaToMove    Replica to move away from its current broker.
-   * @param numLocalReplicas Number of local replicas.
    * @param optimizedGoals   Goals that have already been optimized. The function ensures that their requirements won't
    *                         be violated.
    */
-  void moveSelfHealingEligibleReplicaToEligibleBroker(ClusterModel clusterModel,
-                                                      Replica replicaToMove,
-                                                      int numLocalReplicas,
-                                                      Set<Goal> optimizedGoals)
+  void moveSelfHealingEligibleReplicaToEligibleBroker(ClusterModel clusterModel, Replica replicaToMove, Set<Goal> optimizedGoals)
       throws OptimizationFailureException {
+    int numLocalReplicas = replicaToMove.broker().replicasOfTopicInBroker(replicaToMove.topicPartition().topic()).size();
     // Sanity check the number of replicas to move from the local to a remote broker to achieve the distribution
     // target. If the broker is dead, the replica must move to another broker.
-    if (replicaToMove.broker().isAlive() && numReplicasToMove(numLocalReplicas) == 0) {
+    if (!replicaToMove.isCurrentOffline() && numReplicasToMove(numLocalReplicas) == 0) {
       return;
     }
 
     // Attempt to move the replica to an eligible broker.
     boolean isMoveSuccessful = moveReplicaToEligibleBroker(clusterModel, replicaToMove, optimizedGoals);
-    if (!replicaToMove.broker().isAlive()) {
-      throw new OptimizationFailureException("Self healing failed to move the replica away from decommissioned broker.");
+    if (replicaToMove.isCurrentOffline()) {
+      throw new OptimizationFailureException("Self healing failed to move the offline replica away from broker.");
     }
 
     // Update consumed warm broker credits. Credit is consumed because the broker was unable to move replicas.
```

Comment:
in line 135, `isMoveSuccessful` ->`!isMoveSuccessful`

Reply:
Can you clarify why do you think it should be `!isMoveSuccessful`?

Reply:
Sorry for not elaborating the reason.
we want to move this replica because this broker can more than average replica(thus should consumer warm broker credit), but if we move this replica off this broker,  then this broker's replica count decrease and should not consume warm broker credit(because of this replica).
I think the comments before this check also express the same idea.(https://github.com/linkedin/cruise-control/blob/dbb8607c73bc402dd604b2f45aafead396f7cf0e/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaDistributionTarget.java#L134)